### ISSUE TITLE
verifyTransfer returns false if all modules return Results.NA

### ISSUE THREAD

**Issue by ghost** [PolymathNetwork/polymath-core#411]
The latest verifyTransfer function in SecurityToken.sol is as follows : 
```
    function _verifyTransfer(
        address _from,
        address _to,
        uint256 _value,
        bytes _data,
        bool _isTransfer
    ) internal checkGranularity(_value) returns (bool) {
        if (!transfersFrozen) {
            bool isInvalid = false;
            bool isValid = false;
            bool isForceValid = false;
            bool unarchived = false;
            address module;
            for (uint256 i = 0; i < modules[TRANSFER_KEY].length; i++) {
                module = modules[TRANSFER_KEY][i];
                if (!modulesToData[module].isArchived) {
                    unarchived = true;
                    ITransferManager.Result valid = ITransferManager(module).verifyTransfer(_from, _to, _value, _data, _isTransfer);
                    if (valid == ITransferManager.Result.INVALID) {
                        isInvalid = true;
                    } else if (valid == ITransferManager.Result.VALID) {
                        isValid = true;
                    } else if (valid == ITransferManager.Result.FORCE_VALID) {
                        isForceValid = true;
                    }
                }
            }
            // If no unarchived modules, return true by default
            return unarchived ? (isForceValid ? true : (isInvalid ? false : isValid)) : true;
        }
        return false;
    }
```

I am not sure if I am missing anything out in my understanding, however if here we assume checkGranularity passes and transfers are not frozen, then based on the code if all the modules attached to the security token return are all unarchived and return Result.NA, then 
```
return unarchived ? (isForceValid ? true : (isInvalid ? false : isValid)) : true;
```
will return false, which would make the transfer fail. 

Would this scenario ever happen, or did i misunderstand something?

Thanks.

**Comment by adamdossa** [comment]
Hi @xAffinity - thanks for diving into the code!

The behaviour you described is intentional. If there are TransferManager modules attached to the token, we require at least one module to give a positive response in order for the transfer to be valid.

As an example, the GeneralTransferManager will return VALID if an address appears on it's whitelist, but NA if an address doesn't appear on the whitelist. Another module (e.g. the ManualApprovalTransferManager) may then return VALID (if a manual approval has been set up for this specific transfer) and so taken together the transfer is valid (i.e. NA + VALID = VALID). If all modules return NA then it means no module thinks the transfer is VALID, and hence it is rejected.

If there are no modules attached (or they are all archived) then we consider that the issuer has set the token to have no transfer restrictions, and hence all transfers are approved (assuming it isn't frozen etc.).

Let me know if the above isn't clear.

**Comment by ghost** [comment]
Hi @adamdossa , thanks for your detailed explanation.

You mentioned in your response:
> If there are TransferManager modules attached to the token, we require at least one module to give a positive response in order for the transfer to be valid.

However, in the verifyTransfer method in CountTransferMethod.sol below : 
```
   function verifyTransfer(
        address /* _from */,
        address _to,
        uint256 /* _amount */,
        bytes /* _data */,
        bool /* _isTransfer */
    )
        public
        returns(Result)
    {
        if (!paused) {
            if (maxHolderCount < ISecurityToken(securityToken).getInvestorCount()) {
                // Allow transfers to existing maxHolders
                if (ISecurityToken(securityToken).balanceOf(_to) != 0) {
                    return Result.NA;
                }
                return Result.INVALID;
            }
            return Result.NA;
        }
        return Result.NA;
    }
```
CountTransferManager never returns Result.VALID, so in a case it is the only module or with other modules which all return NA, and hence would make transfers fail even if they should respect the CountTransfer rules.

I understand you changed certain TransferManagers  ( CountTransfer, PercentageTransfer) from having a positive validation return Result.VALID to Result.NA to fix #113 . However, I am not sure if this would break the logic as quoted from you above. 

As an example, if i want to create a SecurityToken with just CountTransfer or PercentageTransfer, the transfer will never be valid since Result.VALID will never be returned.

Please correct me if I am wrong.

Thanks!

**Comment by ghost** [comment]
@adamdossa 

I think I see it now. The logic implemented assumes that by default GTM is attached to the SecurityToken(as expected from comments in STFactory.sol) , then any verifyTransfer will return Result.VALID if allowAllTransfers is true, and the addresses are on the whitelist and satisfy the TimeRestrictions.  So the GTM is essentialy a 'key module' among all modules to ensure basic whitelist requirements of the participants of a security token transaction have to be met in order for the transfer to be valid(even if the transfer respects all other modules and they all return Result.NA) . 

Thanks!
